drivers/i2c/i2c_gd32.c

```diff
@@ -0,0 +1,686 @@
+#ifdef I2C_FMPCFG
+	case I2C_SPEED_FAST_PLUS:
+		if (freq < I2CCLK_FM_PLUS_MIN) {
+			LOG_ERR("I2C fast-mode plus min clock freq %u, current is %u\n",
+				I2CCLK_FM_PLUS_MIN, freq);
+			err = -ENOTSUP;
+			goto error;
+		}
+
+		/* Fast-mode plus risetime maximum value: 120ns */
+		I2C_RT(cfg->reg) = freq * 120U / 1000U + 1U;
+
+		/* pclk1 = bitrate * (clkc * (t_high_scale_9 + t_low_scale_16)) */
+		clkc = pclk1 / (I2C_BITRATE_FAST_PLUS * 25);
+		if (clkc == 0U) {
+			clkc = 1U;
+		}
+
+		/* Default DCTY to 1 */
+		I2C_CKCFG(cfg->reg) |= I2C_CKCFG_DTCY;
+		/* Clear and Set clkc */
+		I2C_CKCFG(cfg->reg) &= ~I2C_CKCFG_CLKC;
+		I2C_CKCFG(cfg->reg) |= clkc;
+		/* Transfer mode: fast-mode */
+		I2C_CKCFG(cfg->reg) |= I2C_CKCFG_FAST;
+
+		/* Enable transfer mode: fast-mode plus */
+		I2C_FMPCFG(cfg->reg) |= I2C_FMPCFG_FMPEN;
+
+		break;
+#endif /* I2C_FMPCFG */
+	default:
+		err = -EINVAL;
+		goto error;
+	}
+
+	data->dev_config = dev_config;
+error:
+	k_sem_give(&data->bus_mutex);
+
+	return err;
+}
+
+static struct i2c_driver_api i2c_gd32_driver_api = {
+	.configure = i2c_gd32_configure,
+	.transfer = i2c_gd32_transfer,
+};
+
+static int i2c_gd32_init(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+	uint32_t bitrate_cfg;
+	int err;
+
+	err = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Mutex semaphore to protect the i2c api in multi-thread env. */
+	k_sem_init(&data->bus_mutex, 1, 1);
+
+	/* Sync semaphore to sync i2c state between isr and transfer api */
+	k_sem_init(&data->sync_sem, 0, K_SEM_MAX_LIMIT);
+
+	rcu_periph_clock_enable(cfg->rcu_periph_clock);
+
+	cfg->irq_cfg_func();
+
+	bitrate_cfg = i2c_map_dt_bitrate(cfg->bitrate);
+
+	i2c_gd32_configure(dev, I2C_MODE_MASTER | bitrate_cfg);
+
+	return 0;
+}
+
+#define I2C_GD32_INIT(inst)							\
+	PINCTRL_DT_INST_DEFINE(inst)						\
+	static void i2c_gd32_irq_cfg_func_##inst(void)				\
+	{									\
+		IRQ_CONNECT(DT_INST_IRQ_BY_NAME(inst, event, irq),		\
+			    DT_INST_IRQ_BY_NAME(inst, event, priority),		\
+			    i2c_gd32_event_isr,					\
+			    DEVICE_DT_INST_GET(inst),				\
+			    0);							\
+		irq_enable(DT_INST_IRQ_BY_NAME(inst, event, irq));		\
+										\
+		IRQ_CONNECT(DT_INST_IRQ_BY_NAME(inst, error, irq),		\
+			    DT_INST_IRQ_BY_NAME(inst, error, priority),		\
+			    i2c_gd32_error_isr,					\
+			    DEVICE_DT_INST_GET(inst),				\
+			    0);							\
+		irq_enable(DT_INST_IRQ_BY_NAME(inst, error, irq));		\
+	}									\
+	static struct i2c_gd32_data i2c_gd32_data_##inst;			\
+	const static struct i2c_gd32_config i2c_gd32_cfg_##inst = {		\
+		.reg = DT_INST_REG_ADDR(inst),					\
+		.bitrate = DT_INST_PROP(inst, clock_frequency),			\
+		.rcu_periph_clock = DT_INST_PROP(inst, rcu_periph_clock),	\
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),			\
+		.irq_cfg_func = i2c_gd32_irq_cfg_func_##inst,			\
+	};									\
+	DEVICE_DT_INST_DEFINE(inst,						\
```

Comment:
This should use the newer I2C_DT_INST_DEFINE https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/i2c.h#L493 rather than the generic DEVICE_DT_INST_DEFINE

Reply:
thanks for spot it. done.